src/main/java/frc/robot/subsystems/drive/SwerveModule.java

```diff
@@ -0,0 +1,237 @@
+   *
+   * @return The current state of the module.
+   */
+  public SwerveModuleState getState() {
+    // Apply chassis angular offset to the encoder position to get the position
+    // relative to the chassis.
+    return new SwerveModuleState(
+        drivingTalon.getVelocity().getValue()*DriveConstants.DRIVING_ENCODER_POSITION_FACTOR_METERS,
+        new Rotation2d(turningEncoder.getPosition() - chassisAngularOffsetRadians));
+  }
+
+  /**
+   * Returns the current position of the module.
+   *
+   * @return The current position of the module.
+   */
+  public SwerveModulePosition getPosition() {
+    // Apply chassis angular offset to the encoder position to get the position
+    // relative to the chassis.
+    return new SwerveModulePosition(
+        drivingTalon.getPosition().getValue()*DriveConstants.DRIVING_ENCODER_POSITION_FACTOR_METERS, 
+        new Rotation2d(turningEncoder.getPosition() - chassisAngularOffsetRadians));
+  }
+
+  /**
+   * Sets the desired state for the module.
+   *
+   * @param desiredState Desired state with speed and angle.
+   */
+  public void setDesiredState(SwerveModuleState desiredState) {
+    // Apply chassis angular offset to the desired state.
+    SwerveModuleState correctedDesiredState = new SwerveModuleState();
+    correctedDesiredState.speedMetersPerSecond = desiredState.speedMetersPerSecond;
+    correctedDesiredState.angle =
+        desiredState.angle.plus(Rotation2d.fromRadians(chassisAngularOffsetRadians));
+
+    // Optimize the reference state to avoid spinning further than 90 degrees.
+    SwerveModuleState optimizedDesiredState =
+        SwerveModuleState.optimize(
+            correctedDesiredState, new Rotation2d(turningEncoder.getPosition()));
+
+    // Setting global desiredState to be optimized for the shuffleboard
+    this.desiredState = optimizedDesiredState;
+
+    // Command driving and turning SPARKS MAX towards their respective setpoints.
+    drivingTalon.setControl(
+        new DutyCycleOut(optimizedDesiredState.speedMetersPerSecond));
+    turningPIDController.setReference(
+        optimizedDesiredState.angle.getRadians(), CANSparkMax.ControlType.kPosition);
+  }
+
+  /** Zeroes all the SwerveModule encoders. */
+  public void resetDriveEncoder() {
+    drivingTalon.setPosition(0);
+  }
+
+  public double getEncoderAbsPositionRad() {
+    return turningEncoder.getPosition();
+  }
+
+  public void periodic() {
+    turningAbsoluteEncoderChecker.addReading(turningEncoder.getPosition());
+  }
+
+  public void initSendable(SendableBuilder builder) {
+    builder.addDoubleProperty("Driving kP", ()->{return drivingTalon.getClosedLoopProportionalOutput().getValue();}, null); //TODO find a setter? i couldn't
+    builder.addDoubleProperty("Driving kI", ()->{return drivingTalon.getClosedLoopIntegratedOutput().getValue();}, null);
+    builder.addDoubleProperty("Driving kD", ()->{return drivingTalon.getClosedLoopDerivativeOutput().getValue();}, null);
```

Comment:
I think this is the contribution to the output from the P term, not the constant?

Reply:
fixed, the above values are now marked as their respective contribution to output from the different terms

Reply:
and the constants are also outputted to the builder